### ISSUE TITLE
Android CMake build on Windows fails

### ISSUE THREAD

**Issue by kurtzmarc** [grpc/grpc#19592]
### What version of gRPC and what language are you using?
1.22 
 
### What operating system (Linux, Windows,...) and version?
 Windows 10.
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 Android cross-compile w/ latest NDK. CMake 3.6. Ninja 1.5.3. Go 1.12.6. Strawberry Perl 5.30.0.1.
 
### What did you do?

1. Checked out master.
1. Brought up CMake-GUI. Point it to `grpc/examples/android/helloworld/app`.
1. Set `CMAKE_MAKE_PROGRAM` to Android NDK's ninja.
1. Set `HAVE_GNU_POSIX_REGEX`, etc... to `0`
1. Set `helloworld_PROTOBUF_PROTOC_EXECUTABLE` to point to latest build of `protoc.exe`
1. Set `helloworld_GPRC_CPP_PLUGIN_EXECUTABLE` to point to latest build of `grpc_cpp_plugin.exe`
1. Configure. Generate.
1. `cd <outputdir>`, `<android ndk path>\ninja.exe`.

### What did you expect to see?
Success!
 
### What did you see instead?
Failure : ( 

```
[2026/2070] Building CXX object outputs/armeabi-v7a/CMakeF.../src/proto/grpc/reflection/v1alpha/reflection.grpc.pb.cc.o
FAILED: C:\Android\sdk\ndk-bundle\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe --target=armv7-none-linux-androideabi16 --gcc-toolchain=C:/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64 --sysroot=C:/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64/sysroot  -DCARES_STATICLIB -DPB_FIELD_32BIT -I../../third_party/zlib -I../../include -I../../ -I../../third_party/boringssl/include -I../../third_party/protobuf -Ioutputs/armeabi-v7a/third_party/zlib -I../../third_party/benchmark/include -I../../third_party/cares/cares -Ioutputs/armeabi-v7a/third_party/cares/cares -Ioutputs/armeabi-v7a/third_party/gflags/include -I../../third_party/address_sorting/include -I../../third_party/nanopb -Igens -I../../third_party/protobuf/src -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -fno-addrsig -march=armv7-a -mthumb -Wa,--noexecstack -Wformat -Werror=format-security   -std=c++11 -std=c++11 -fPIC -MD -MT outputs/armeabi-v7a/CMakeFiles/grpc++_reflection.dir/outputs/grpc/gens/src/proto/grpc/reflection/v1alpha/reflection.grpc.pb.cc.o -MF outputs\armeabi-v7a\CMakeFiles\grpc++_reflection.dir\outputs\grpc\gens\src\proto\grpc\reflection\v1alpha\reflection.grpc.pb.cc.o.d -o outputs/armeabi-v7a/CMakeFiles/grpc++_reflection.dir/outputs/grpc/gens/src/proto/grpc/reflection/v1alpha/reflection.grpc.pb.cc.o -c gens/src/proto/grpc/reflection/v1alpha/reflection.grpc.pb.cc
In file included from gens/src/proto/grpc/reflection/v1alpha/reflection.grpc.pb.cc:6:
gens\src/proto/grpc/reflection/v1alpha/reflection.grpc.pb.h:162:13: error: no member named 'internal' in namespace 'grpc_impl'; did you mean simply 'internal'?
        new ::grpc_impl::internal::CallbackBidiHandler< ::grpc::reflection::v1alpha::ServerReflectionRequest, ::grpc::reflection::v1alpha::ServerReflectionResponse>(
            ^~~~~~~~~~~~~~~~~~~~~
            internal
../../include\grpcpp/impl/codegen/proto_buffer_writer.h:40:11: note: 'internal' declared here
namespace internal {
          ^
```
 
### Anything else we should know about your project / environment?

I suspect your CI builds are on Linux so this is likely Windows specific when cross compiling for Android. I reviewed #8398 and #10681 and it appears Windows cross compiling was working at some point. I'd like to help get that working again!

**Comment by jtattermusch** [comment]
Have you tried using gradle to build?
https://github.com/grpc/grpc/tree/master/src/android/test/interop#install

**Comment by jtattermusch** [comment]
FTR, the entrypoint for linux android build is here: https://github.com/grpc/grpc/blob/master/tools/internal_ci/linux/grpc_android.sh
https://github.com/grpc/grpc/tree/master/examples/android/helloworld

**Comment by kurtzmarc** [comment]
Thank you for pointing me to that. I had to do a few extra steps:
1. Update the gradle version in `gradle/wrapper/gradle-wrapper.properties` to `distributionUrl=https\://services.gradle.org/distributions/gradle-4.6-all.zip`
1. Update my gradle plugin in `build.gradle` to `classpath 'com.android.tools.build:gradle:3.2.0'`
1. Add some cmake arguments to point to my Go and Perl executables:
    ```
    arguments '-DGO_EXECUTABLE=C:/go/bin/go.exe'
    arguments '-DPERL_EXECUTABLE=C:/perl/perl/bin/perl.exe'
    ```
1. I had to build on my development station protoc.exe and grpc_cpp_plugin.exe and copy them over to the interop directory. Then point gradle.properties to them:
    ```
    protoc=..\\protoc.exe
    grpc_cpp_plugin=..\\grpc_cpp_plugin.exe
    ```
After that it would successfully build using gradle. That would produce an apk, however the build artifacts which contain the libraries are found in `app\build\intermediates\cmake`.

Additionally, if I use the flags set in gradle.build in the cmake-gui (or command line) I can build the libraries outside of gradle using cmake/ninja against the Android toolchain.
